### DISCUSSION TITLE
Talented codes: SDK import tutorials needed

### DISCUSSION THREAD

**Discussion by jake653** [earlephilhower/arduino-pico#156] [Q&A]
Talented Coders:

Needed by the less talented: a tutorial on importing SDK source files into Arduino.

Those with little professional tool chain experience 
who rely on example/template files as a jumping off point 
have quite a challenge at this stage of development.
(It would seem that converting a rp2040 optimized SDK project
would be more expedient that starting with an old Arduino project)

- - - - 

My experience:
(1)
Got a simple single source file example working by removing  stdio_init_all();
  (taking more time than it should have)

- - - - 
 (2)
Official SDK PIO example files don't include .pio.h files, for example:

blink.c  contains #include "blink.pio.h"  but source files only have "blink.pio"

https://github.com/raspberrypi/pico-examples/tree/master/pio/pio_blink

While Earl's  Tone.cpp

 arduino-pico/cores/rp2040/Tone.cpp  contains:  #include "tone.pio.h"

Has both tone.pio & tone.pio.h files

tone.pio.h contains:

// -------------------------------------------------- //
// This file is autogenerated by pioasm; do not edit! //
// -------------------------------------------------- //

Is pioasm needed to run SDK examples in Arduino?

- - - -
(3)
External file organization/naming

Consulted  pre-processing documentation and discussions.
https://arduino.github.io/arduino-cli/latest/sketch-build-process/#pre-processing

It appears that external .c and .h files are treated as C code while .ino are converted to .cpp

In an example for the SDK that includes a few old Arduino .c & .h files:

Trying to translate a Nokia LCD for Pi Pico example, that used a couple external .c and .h files,
Errors reported for:  Documents/Arduino/Nokia_5110/Nokia_5110.ino:

 warning: implicit declaration of function 'abs' [-Wimplicit-function-declaration]
   bool steep = (abs(y1 - y0) > abs(x1 - x0));   // guessing it's a C and C++ type mixup?

got these error messages for functions contained in external files:

undefined reference to `_Z8LCD_Demov'         (.ino src: LCD_Demo();)
undefined reference to `_Z12clearDisplayv'    (.ino src: clearDisplay();)  
undefined reference to `_Z14Nokia5110_Initv'  (.ino src: Nokia5110_Init();)

Guessing the odd Znn... must be intermediate process versions?

I tried  "C" {} block marking on external handler calls, getting same errors with no effect.

**Comment by earlephilhower** [comment]
(1) You can get some things running (the simple SDK examples) by replacing main() with setup() and removing some unneeded bits like you said, but honestly if you have a SDK project, the default `cmake` workflow may be better for you than trying to move it to Arduino.  The gotcha is the SDK could change the machine state in ways the Arduino libraries don't know about (i.e. that serial USB bit would re-init the USB hardware and make bad stuff happen).

(2) You need to manually run `pioasm` to compile the `.pio` to `.pio.h`.  It's not part of the IDE because Arduino doesn't support generic makefile targets.  I include `pioasm` as one of the installed tools, though, so just look in the installed tools dir to find it.

(3) The `abs` thing is a bug in my `Arduino.h`.  It's kind of a mess in the generic Arduino core (it redefines a standard C function in an unsafe, incompatible way), and it's always a point of contention in new cores.  I should define it one way or the other, at least.

The _Z... labels are C++ name mangled functions that aren't present in your link for some reason.  See https://demangler.com/ to unscramble them.

**Reply by jake653** [reply]
Thanks for taking the time...
I don't see pioasm in the Arduino 1.8.13 Tools menu on my OSX system, with your release 1.5.1 installed.

TRIVIA I did see some inline piasm examples for circuitpython:
pioasm_i2sout.py,  pioasm_neopixel.py,  pioasm_pdm.py, pioasm_simpletest.py (.program squarewave)

- - - - - 
About the mangled handler name referenced in error messages:
Not sure about that demangler.com reference.

Ii I renamed the external file to .cpp instead of .c  the handler name in the error msg was not mangled.  (originally a .c)
 (but got new ISO C++ string violations)

  The Nokia5110 for pico source file I was trying to translate, sorry...
 actually one of 11  "Example Codes for Raspberry Pi Pico with C/C++ SDK using VS Code"
https://github.com/viduraembedded/Raspberry-Pi-Pico
https://www.youtube.com/watch?v=Ll23kHzQrtA
- - - - - 

The abs() error went away when test compiling a code fragment in a primary .ino instead of an external file.
    bool steep = (abs(y1 - y0) > abs(x1 - x0)); // was OK

**Reply by earlephilhower** [reply]
You need to run `pioasm` from the command line.  It's a CLI utility, not something to run from the IDE.

The name mangling stuff makes sense now.  You're mixing C and CPP files with a header that doesn't specify the right name mangling option.  Easiest solution, make everything .CPP and fix the warnings that come up (C++ is more pedantic and catches more errors than C will at the compile stage).  OTW, look at some of the headers in the repo who have the `extern "C"` command in them to learn about ensuring C linkage of functions.

Good luck!
